moveit_core/kinematics_base/include/moveit/kinematics_base/kinematics_base.h

```diff
@@ -612,34 +615,55 @@ class KinematicsBase
   template <typename T>
   inline bool lookupParam(const std::string& param, T& val, const T& default_val) const
   {
-    ros::NodeHandle pnh("~");
-    if (pnh.hasParam(group_name_ + "/" + param))
-    {
-      val = pnh.param(group_name_ + "/" + param, default_val);
-      return true;
-    }
+    auto node = rclcpp::Node::make_shared("lookup_param");
 
-    if (pnh.hasParam(param))
+    auto parameters_lookup = std::make_shared<rclcpp::SyncParametersClient>(node);
+
+    auto groupname_param = parameters_lookup->get_parameters({ group_name_ + "/" + param });
```

Comment:
`auto` is nice where the type is obvious but in cases like this it's more confusing then helpful.
Especially since the following for loops are using `auto` as well this type should be made explicit.

Reply:
Also, the `SyncParametersClient` class contains the functions `has_param()` and `get_parameter()` which can be used to replace `hasParam()` and `param()` directly. This way we don't have to loop and compare the results as well.

Reply:
Partially complied with your request @henningkayser at b2d436f024d1003ea57d9f8f6ff88ddde0f3c9a1.

Please make a suggestion if you require further changes.

Reply:
`get_parameters()` is meant to be used for multiple parameters but here it's always called with only a single entry. In this case it's much more straightforward to query every parameter like in the sample code below:
```C++
    bool param_success = node.get_parameter<T>(group_name + "/" + param, val);
```

No need to create a parameter vector or loop over the results and also we can use the node directly and don't need to cast it to `rclcpp::SyncParametersClient`.

Since `lookupParam()` does in fact query multiple parameters in different locations we can absolutely use `Node::get_parameters()` but in that case including all parameter names:
```C++
val = default_val;
std::vector<std::string> param_names = {
  group_name_ + "/" + param, param,
  "robot_description_kinematics/" + group_name_ + "/" + param,
  robot_description_kinematics/" + param
};
std::vector<rclcpp::parameter::ParameterVariant> param_results =
  node.getParameters(param_names);
bool param_success = !param_results.empty();
if (param_success)
    val = param_results[0].get_value<T>();
return param_success;
```

I actually prefer the second implementation.
Please also note that your current implementation calls `value_to_string()` on all parameters but `lookupParam()` is a template function that supports all parameter types. The `get_value()` function takes care of using the correct type.

Reply:
I've already made some changes (on planning_scene_monitor) related to this:

https://github.com/AcutronicRobotics/moveit2/blob/5a8ad535e33548434b02e5e9e735ab614465f5a5/moveit_ros/planning/planning_scene_monitor/src/planning_scene_monitor.cpp#L1472-L1516

Reply:
@anasarrak this looks much better. However, each if/else block can be implemented in a single line using `node->get_parameter_or(param, val, default_val)`.